### PR TITLE
block/aoe: use BPF to filter requests for other servers

### PR DESCRIPTION
**Pull request by mdlayher**
I added functionality tonight to my raw sockets package to enable attaching a BPF program to the raw socket.  `glide` gave me some difficulties so that package bump is not included here.  I'd be happy to look into getting that done tomorrow.

This is the improvement I mentioned in #184 , and should stop the AoE server from getting bogged down handling requests that it doesn't care about.

### CODE REVIEW

**Comment by mischief** [comment]
what happens if the mtu changes at runtime?

**Comment by mdlayher** [comment]
The interface's MTU is already relied on for:
- allocating a byte slice to read in ethernet frames
- determining sector size to report to AoE clients

So I imagine this would not be the only problem area if the interface MTU were to change underneath the service.  In this case, we could also just hardcode it to 9000 since the return value just asks BPF to return "up to X" bytes.

File: block/aoe/aoe.go [hunk 1]
```diff
@@ -211,3 +210,63 @@ func (s *Server) handleFrame(from net.Addr, iface *Interface, f *Frame) (int, er
...
 	return s.dev.Close()
 }
+
+// mustAssembleBPF assembles a BPF program to filter out packets not bound
+// for this server.
+func (s *Server) mustAssembleBPF(mtu int) []bpf.RawInstruction {
+	prog, err := bpf.Assemble([]bpf.Instruction{
+		// Load major address value from AoE header
+		bpf.LoadAbsolute{
+			Off:  16,
```
Comment (xiang90) [inline block/aoe/aoe.go]: i hope there is a const for location/size of the major field in header.  but seems like the aoe pkg does not expose it. :(

Comment (mdlayher) [inline block/aoe/aoe.go]: Yeah, I tried to keep the API surface small for that package if at all possible.  Perhaps this BPF program should reside in the AoE package instead.  I do want to create a proper "server" type in that package which would definitely overlap with some of the code in this package.

For the time being, I could add comments explaining the meaning of this magic number, if that works.

Comment (xiang90) [inline block/aoe/aoe.go]: Sure. I think a few comments should be good enough for now.

**Comment by mdlayher** [comment]
Updated, @xiang90 !  @mischief , any more thoughts on the MTU issue?

**Comment by barakmich** [comment]
All looks good to me operationally. I'd say, yeah, you probably want a server type, of which this is roughly a good first implementation. Someday, when you want to bake it into your aoe package, we'll just use it :)

:+1:

**Pull request merged**